implementations/elixir/ockam/ockam/lib/ockam/transport/portal/interceptor.ex

```diff
@@ -0,0 +1,274 @@
+
+  @doc """
+  Process intercepted payload from outer worker.
+  Returns list of payloads to forward, which can be used to buffer messages of different size.
+  This is done because tcp portal batches tcp packets.
+  Returns:
+   - {:ok, payloads, state} - forward 0 or more payloads
+   - {:stop, reason, payloads, state} - forward 0 or more payloads and terminate the interceptor
+  """
+  @callback handle_outer_payload(payload :: binary(), state :: map()) ::
+              {:ok, payloads :: [binary()], state :: map()}
+              | {:stop, reason :: any(), payloads :: [binary()], state :: map()}
+  @doc """
+  Process intercepted payload from inner worker.
+  Returns list of payloads to forward, which can be used to buffer messages of different size.
+  This is done because tcp portal batches tcp packets.
+  Returns:
+   - {:ok, payloads, state} - forward 0 or more payloads
+   - {:stop, reason, payloads, state} - forward 0 or more payloads and terminate the interceptor
+  """
+  @callback handle_inner_payload(payload :: binary(), state :: map()) ::
+              {:ok, payloads :: [binary()], state :: map()}
+              | {:stop, reason :: any(), payload :: [binary()], state :: map()}
+  @doc """
+  Process intercepted signal from outer worker (:ping, :pong, :disconnect).
+  Returns:
+   - {:ok, state} - will forward original signal
+   - {:error, reason} - will not forward the signal
+  """
+  @callback handle_outer_signal(signal :: any(), state :: map()) ::
+              {:ok, state :: map()}
+              | {:stop, reason :: any(), state :: map()}
+  @doc """
+  Process intercepted signal from inner worker (:ping, :pong, :disconnect).
+  Returns:
+   - {:ok, state} - will forward original signal
+   - {:error, reason} - will not forward the signal
+  """
+  @callback handle_inner_signal(signal :: any(), state :: map()) ::
+              {:ok, state :: map()}
+              | {:stop, reason :: any(), state :: map()}
+
+  @impl true
+  def inner_setup(options, state) do
+    interceptor_mod = Keyword.fetch!(options, :interceptor_mod)
+    interceptor_options = Keyword.get(options, :interceptor_options, [])
+
+    case interceptor_mod.setup(interceptor_options, state) do
+      {:ok, state} ->
+        state = Map.put(state, :interceptor_mod, interceptor_mod)
+
+        case Keyword.fetch(options, :init_message) do
+          {:ok, message} ->
+            ## Interceptor is spawned by Ockam.Session.Spawner
+            handle_outer_message(message, state)
+
+          :error ->
+            {:ok, state}
+        end
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  @impl true
+  def handle_outer_message(message, state) do
+    handle_message(:outer, message, state)
+  end
+
+  @impl true
+  def handle_inner_message(message, state) do
+    handle_message(:inner, message, state)
+  end
+
+  def handle_message(direction, %Message{} = message, state) do
```

Comment:
Just a general comment - I'm curious how do we decide here what are public vs private functions. For example in this module most of the public functions (or even all outside of callback implementations) are internal to the module. Could we make them private, so it's clear that these are not APIs?

Reply:
Generally it's better to make functions private unless they need to be public (for API or testing reasons), but I tend to forget.